New Coupons Spice from www.sparheld.de

ingowachsmuth-sparheld opened the pull request:
Commit Information according to pull request template:

**What does your Instant Answer do?**
This IA searches sparheld.de if coupons and deals exist for a certain shop.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
Users directly see the available coupons and deals for a shop in the search results.

What is the data source for your Instant Answer? (Provide a link if possible)
sparheld.de

**Why did you choose this data source?**
We're not only aggregating coupon codes for online stores from affiliate networks. We get our coupons for example from the following sources :
* Affiliate Networks
* Directly from the Online Stores (exklusives)
* Newsletters (We subscribe to every online stores newsletter. When there's a coupon we list it.)
* Facebook (We like every online stores facebook page. When there's a coupon we list it.)
* Competitor (In the end, we also check other coupon sites. If they have an additional coupon we add it.)

All coupon codes which go on one of our sites are always tested manually by an employee before they go live. If they work, we add it. If they don't we, of course, don't add it. Furthermore, we continously test the coupon codes after they're online as we want to ensure that all coupon codes on our sites work.

**Are there any other alternative (better) data sources?**
No

**What are some example queries that trigger this Instant Answer?**
Zooplus Gutschein
Gutschein H&M

Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)
online shoppers

**Is this Instant Answer connected to a DuckDuckHack Instant Answer idea?**
Yes

**Which existing Instant Answers will this one supersede/overlap with?**
None

**Are you having any problems? Do you need our help with anything?**
Yes. Is it possible to set the triggers, API-urls, website name etc. all by locale? In the end of the day, for each country, we have a different brand name and API-urls are also different. If at all possible, we would like to have one "spice" where we make differences by locale. We wouldn't prefer to do one spice per country as it's long-term more maintenance for all of us. Then, we can also provide online-shoppers in 7 more countries with coupons and deals as IA.

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
Free of Charge. API was a custom development for DuckDuckGo. DuckDuckGo got a specific API-Key for authorization.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
http://i.imgur.com/6YuaQVH.png

Future Ideas
After discussion by e-mail with DuckDuckGo-Team-Member, we used the media template. Long-term, if there's a new template, we could do something like this which would be possibly better for the user:
https://cloud.githubusercontent.com/assets/7061920/3544158/ade4359a-0869-11e4-930e-98e02505dde6.png

```
COMMIT CHECKLIST:
[X] Added metadata and attribution information
[X] Wrote test file and added to t/ directory
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[as good as possible] Tested cross-browser compatibility


Please let us know which browsers/devices you've tested on:

- Windows 10
    - [ ] Google Chrome
    - [ ] Firefox
    - [ ] Opera
    - [ ] IE 11
    - [ ] Microsoft edge

- Windows 8
    - [ ] Google Chrome
    - [ ] Firefox
    - [ ] Opera
    - [ ] IE 10

- Windows 7
    - [X] Google Chrome
    - [X] Firefox
    - [ ] Opera
    - [ ] IE 8
    - [ ] IE 9
    - [ ] IE 10

- Windows XP
    - [ ] IE 7
    - [ ] IE 8

- Linux
    - [X] Google Chrome
    - [X] Firefox

- Mac OSX
    - [ ] Google Chrome
    - [ ] Firefox
    - [ ] Opera
    - [X] Safari

- iOS (iPhone)
    - [ ] Safari Mobile
    - [ ] Google Chrome
    - [ ] Opera

- iOS (iPad)
    - [ ] Safari Mobile
    - [ ] Google Chrome
    - [ ] Opera

- Android
    - [ ] Firefox
    - [X] Native Browser
    - [ ] Google Chrome
    - [ ] Opera
```
-----
IA Page: https://duck.co/ia/view/coupons

daxtheduck commented:
@ingowachsmuth-sparheld Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

While the team and the community review your submission, we'd like to make sure you look good! If you'd like a free DuckDuckGo tshirt, please send over the following info to merch@duckduckgo.com, and be sure to put "Dev Merch" in the subject line:

Name:
Street Address:
City:
State / Province:
Country:
Phone (Non-US Orders only): 
T-shirt size:
GitHub ID:
Link to this Pull Request:

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please send an email to <quackslack@duckduckgo.com>.

>More Info: https://duck.co/duckduckhack/submission_and_review

Thanks!

moollaza commented:
@ingowachsmuth-sparheld thanks for this!

>Is it possible to set the triggers, API-urls, website name etc. all by locale?

Triggers aren't regional, so this shouldn't matter, we can create a file with all your triggers, for all languages, and import them.

You can read about loading Data Files here: https://talsraviv.gitbooks.io/duckduckhackdocs/content/duckduckhack/backend-reference/data-files.html (note this is an experimental version of our documentation!)

Here's a live example: https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/Spice/Quandl/HomeValues.pm#L39 

With regards to the API urls, can you give me examples of how they will look?

The `handle` function can return multiple items, which you can capture and inject into the `spice to` URL. We could do something like this for example"

```perl
spice to => 'http://api.sparheld.de/partnerAPI/v1?locale=$2format=json&dataKey=coupons&queryString=$1';

handle query_lc => {
    return unless $loc and $loc->country;
    my $locale = $loc->country;
 
    return $query, $locale
```

If the whole domain is different, then don't have a solution for that right now, besides using several Spice packages. I'll be discussing this with the team to see if we can find an easy solution

ingowachsmuth-sparheld commented:
We added support for different countries via location api.

TravisCI should work now, although I have no possibility to test it myself. At least all required metadata should be there now.

If you would like anything to be changed, just comment here and we will do it asap!

ingowachsmuth-sparheld commented:
Can you tell me why the travis-ci tests are failing? Which metadata is missing?
I can not test this locally atm, as my system refuses to install the required  Dist::Zilla::Plugin::BuildShareAssets

moollaza commented:
> Can you tell me why the travis-ci tests are failing? Which metadata is missing?
I can not test this locally atm, as my system refuses to install the required Dist::Zilla::Plugin::BuildShareAssets

@ingowachsmuth-sparheld this is expected for now. I'll create a link in the PR description to an Instant Answer Page for this IA. Our backend will detect the link and create an IA Page and pull the required metadata from this PR. The test will pass after that, and I will then be able to deploy this to beta.duckduckgo.com for testing

moollaza commented on share/spice/coupons/coupons.js:
```diff
@@ -0,0 +1,88 @@
+        }
+
+        var tabname, translation_ValidUntil, translation_Unknown;
+        switch (api_result.locale) {
```
I think it would be better to have an object defined outside our callback function, with the keys being the locales. Then we don't need to iterate over all locales because we can just do something like `var locale_data = locales[locale]`

ingowachsmuth-sparheld replied:
thats actually a lot more elegant. done.

moollaza commented on share/spice/coupons/coupons.js:
```diff
@@ -0,0 +1,88 @@
+                translation_ValidUntil = 'Voimassa: ';
+                translation_Unknown = 'tuntematon';
+                break;
+            default:
```
Is this case even possible? If so it seems like we wouldn't want to continue execution and should call `Spice.failed('coupons')`

ingowachsmuth-sparheld replied:
Not its not possible outside of testing as the handle function already filters out all other locales.

Still added a check for additional safety.

moollaza commented on share/spice/coupons/coupons.js:
```diff
@@ -0,0 +1,88 @@
+        var tabname, translation_ValidUntil, translation_Unknown;
+        switch (api_result.locale) {
+            case 'DE':
+                tabname = 'Gutscheine';
```
We'd like to use generic "topics" for our categories, ideally from the list in our docs, or viewable here (on the right): https://duck.co/ia -- I think it would be appropriate to set the tab names as `Answer` because this IA directly answers queries about shopping. 

Our frontend will handle translating the word Answer into whichever local the user has set, so we should'nt need to keep a list of all the different words in the code

ingowachsmuth-sparheld replied:
Removed the tab name so it's set to "Answers".
You're right, this should not confuse anyone, as I expect our IA to be very seldom be displayed along with other IAs using a generic name.

moollaza commented on share/spice/coupons/coupons.js:
```diff
@@ -0,0 +1,88 @@
+            normalize: function(item) {
+                return {
+                    title: item.title,
+                    subtitle: "", 
```
we can simply remove this if it's not needed :+1:

ingowachsmuth-sparheld replied:
Ok. Thx!

moollaza commented on share/spice/coupons/coupons.js:
```diff
@@ -0,0 +1,88 @@
+                return {
+                    title: item.title,
+                    subtitle: "", 
+                    description: item.date && item.date.expires ? translation_ValidUntil + new Date(item.date.expires.timestamp * 1000).toLocaleDateString() : translation_ValidUntil + translation_Unknown ,
```
If you'd like, you can use Moment.js to handle date/time strings.

Here's an example: https://github.com/duckduckgo/zeroclickinfo-spice/blob/41561b186c29121086b320db72c3a7d1afcb7b5a/share/spice/wikinews/wikinews.js#L9

moollaza replied:
You just need to wrap `Spice.add()` inside `DDG.require('moment.js', function(){ ... }` and you'll have access to `moment` inside

ingowachsmuth-sparheld replied:
Beautiful :-)

moollaza commented on share/spice/coupons/coupons.js:
```diff
@@ -0,0 +1,88 @@
+                detail: false,
+                item_detail: false,
+                options: {
+                    content: Spice.coupons.content,
```
Seeing as we don't have a custom sub-template you can remove this line and the `coupons.handlebars` file as well.

ingowachsmuth-sparheld replied:
done!

moollaza commented on lib/DDG/Spice/Coupons.pm:
```diff
@@ -0,0 +1,42 @@
+
+
+# Caching - yes
+spice is_cached => 1;
```
We likely don't want this caching enabled because a search for "gutschein h&m" isn't guarenteed to need the same API call, as it won't always come from the same locale. We want to execut the `handle` each time we get this query, to ensure we get the searcher's Locale and make the correct API call.

ingowachsmuth-sparheld replied:
Ahh... I assumed the caching to be based on the full api url.

Disabled it.

ingowachsmuth-sparheld commented:
@moollaza Thank you for you input Zaahir!
I changed everything you addressed.
If I can done anything more, please tell me. If not I look forward to seeing the Instant Answer in beta!

moollaza commented on share/spice/coupons/coupons.js:
```diff
@@ -0,0 +1,83 @@
+                    return {
+                        title: item.title,
+                        description: item.date && item.date.expires ? locale_data.validUntil + moment(item.date.expires.timestamp * 1000).format('L') : locale_data.validUntil + locale_data.unknown,
+                        url: api_result.result.homeurl,
```
Is there a more specific per-coupon, page they can follow through to that would show the coupon they clicked on?

Or, could we show them the code to use, and send them to the retailer's website?

That would be ideal for the user, because sending them to your homepage seems like a bit of a disconnect

ingowachsmuth-sparheld replied:
Oha... that was a mayor mistake on my side!

Of course the user should be send to the page of this exact coupon, with a button to copy the coupon-code. (I don't know how I overlooked breaking that.)

Fixed it!

moollaza commented on share/spice/coupons/coupons.js:
```diff
@@ -0,0 +1,83 @@
+                data: api_result.items,
+                meta: {
+                    sourceName: api_result.result.sitename,
+                    sourceUrl: api_result.result.homeurl
```
Similarly to the below comment, can we direct them to a search page on your website with the query they entered? Typically the `more at` sends them to a place where they can get more info about their query on the provider's domain.

ingowachsmuth-sparheld replied:
Technically we could, but as our search returns the same results the IA gets or directs directly to the shop-page if only one is found, the user would be none the wiser. So I think the homepage is actually more useful to the user.

moollaza commented on share/spice/coupons/coupons_triggers.txt:
```diff
@@ -0,0 +1,139 @@
+gutschein
```
for editing/maintenance it would probably help to alpha-sort this file

ingowachsmuth-sparheld replied:
done.

ingowachsmuth-sparheld commented:
Hi Zaahir!

Excuse my mistake regarding the out-urls. I fixed it.
the caching of the api response is on again. the trigger file is sorted.

I'll do more extensive testing again, if it's deployed to beta.

Thanks for your review (and patience)

moollaza commented:
@ingowachsmuth-sparheld this is now live on beta: https://beta.duckduckgo.com/?q=gutschein+h%26M&ia=iatesting&iai=3

ingowachsmuth-sparheld commented:
@moollaza Thank you!

After some testing, we we're not content with displaying the expiration date of the coupons and having only the logos as a means to distinguish between different shops. We removed the expiration date and display the shop name in the description now. the translations are also unnecessary now, and have been removed.

Also I moved the URLs for out-links and logos to the API, so there is less need for code maintenance now.

I hope you are okay with and not annoyed by these  changes! 

yours,
Ingo

moollaza commented on share/spice/coupons/coupons_triggers.txt:
```diff
@@ -0,0 +1,140 @@
+codice promo
+codice promozionale
+codice promozione
+codice sconto
```
My colleague suggested you add: "codici buono sconto"

moollaza commented on share/spice/coupons/coupons_triggers.txt:
```diff
@@ -0,0 +1,140 @@
+premiekod
+promootiokoodi
+promotion nummer
+promotionskode
```
My colleague suggest you add: "promozione", "sconto" and "sconti"

moollaza commented on share/spice/coupons/coupons_triggers.txt:
```diff
@@ -0,0 +1,140 @@
+codice promozionale
+codice promozione
+codice sconto
+codice vantaggio
```
My Italian colleague suggests we remove this one as it's a poor english -> italian translation

moollaza commented:
@ingowachsmuth-sparheld thanks for all the updates, I had some colleagues review the triggers and I've suggested a few minor changes.

I'd like to have the design team take a look at this now and see if we can make some minimal changes to achieve an acceptable UI.

@abeyang I was thinking of using the `4line` title variant, and moving the brand from the `description` to the `subtitle.

If we use CSS to give the images `1em` of top margin and force an auto-height, the result looks like this:

![gutschein_shoes_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/11641324/b207b412-9d05-11e5-974d-f790542a1d01.png)

@abeyang any thoughts? Should we head a different direction?

Chris made a coupon design a while ago, but it require a custom template and lots of CSS so I'd like to find a simpler solution for now, if possible.

Right now we're displaying all the info we have available to us. There is also an expiration date for some, but I'm not sure if it's worth showing as everything displayed is valid.

Kruthaup commented:
@moollaza Is there a way to test all the locales on beta.duckduckgo.com? We want to test, if the locale is passed correctly to our API, but so far we haven't figured out how to test it.

Kruthaup commented:
@abeyang @moollaza Anything we can do? We'd love to get this online. We can do the suggested minor changes. However, we'd love to do it, if needed, with the layout update. Also, we'd love to test the locales, if we know how.

moollaza commented:
@ingowachsmuth-sparheld apologies for the delay! 

With regards to locales, you can force the locale in DuckPAN like so: `DDG_TEST_LOCATION=de duckpan server`.

For Beta, I'll be speaking with the team on providing a way for developers to force locales for the purpose of testing. Right now, I've personally been using a free VPN, https://www.tunnelbear.com/ to change my IP to the different regions.

We don't typically deal with non-US, region specific Instant Answers so please forgive the lack of tooling available.

----

With regards to the design, we're leaning towards something that looks similar to our GoWatchIt Spice: https://duckduckgo.com/?q=watch+the+dark+knight&t=ffab&ia=watch

If you center the text, reduce the height, add some margin around the images and reduce the tile width you can have something that looks like this:

![gutschein_shoes_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/11801835/a0a1cbe2-a2b7-11e5-935d-28662db29417.png)


To center the text, you can use the `elClass` property:

```js
templates: {
    group: 'media',
    detail: false,
    item_detail: false,
    options: {
        moreAt: true
    },
    elClass: {
        tileBody: "text-center"
    }
}
```

You'll need to use CSS to modify to adjust the tile height/width and image spacing. Make sure to preface your CSS classes with `.zci--coupons` so it only targets your Instant Answer:

```css
.zci--coupons .tile--c {
    width: 14em;
}

.zci--coupons .tile--media {
    height: 5em;
    width: auto;
    margin: 1em;
}

.zci--coupons .tile--content {
    height: auto;
}

```

/cc @abeyang

moollaza commented:
**Update**: the design team would like the tile to have a 24px border:

<img src="https://cloud.githubusercontent.com/assets/873785/11854200/577425fa-a410-11e5-9d9f-604b00276045.png">

ingowachsmuth-sparheld commented:
@moollaza 

We added the CSS for the design so it looks a little nicer. Thanks for the tipps how to do it!

We also tested the locale support with the german locale in duckpan server. We'll test more locales once on beta, as duckpan only supports us,de,my.

Happy holidays to you and your team! I'll be back on the 6th of January. Henning may be reachable. (And wishes a merry christmas too!)

moollaza commented:
>We added the CSS for the design so it looks a little nicer. Thanks for the tipps how to do it!

Wonderful, thanks you!

>We also tested the locale support with the german locale in duckpan server. We'll test more locales once on beta, as duckpan only supports us,de,my.

Excellent. I will be testing this as well with a VPN.

> Happy holidays to you and your team! I'll be back on the 6th of January. Henning may be reachable. (And wishes a merry christmas too!) 

Happy Holidays to you as well! I will try and get this merged and hopefully live before year's end. If not, I'm hoping to have it life within the first week of 2016 :+1: 

Thanks again for doing excellent work

moollaza commented:
BTW please be sure to send us an email, per the instructions above, so your team can get some shirts :+1:

Kruthaup commented:
@moollaza Tested with proxys in Germany, France and Spain. Looks good! For the others, do not have a proxy to test. Really looking forward to see this live!

T-Shirts we already sent an e-mail. Bill already sent it out Dec. 9th. With Christmas, just takes a while, I guess.

moollaza commented:
>Tested with proxys in Germany, France and Spain. Looks good! For the others, do not have a proxy to test. Really looking forward to see this live!

@ingowachsmuth-sparheld wonderful, I'll test the remaining regions!

> T-Shirts we already sent an e-mail. Bill already sent it out Dec. 9th. With Christmas, just takes a while, I guess. 

Excellent! Enjoy your shirts!

@ingowachsmuth-sparheld one question I have -- would there be any reliable way of removing the brand from the code descriptions? It seems like most of the German results end with `bei <company>` and then below we have the company name repeated. 

If we can find a way to pull those out (regex seems like the only multi-lingual approach?) it would be nice to prevent the repetition. I think it's OK for now though.

moollaza commented:
This LGTM :+1:
